### PR TITLE
Contract Self-Signed gasless poll creation & voting

### PR DESCRIPTION
**Pull request by CedarMist**
This hoists my previous gasless voting PR onto the new design + fixes lots of UX things which made it tedious to use.

Deploy with the gasless voting addition:

    $ pnpm hardhat deploy --network sapphire-localnet --gasless

Uses EIP-712 for both poll creation and poll voting. Aside from creating, voting and closing polls there are no other interjections from metamask. Users don't need to have gas on Sapphire to perform these operations.

image

TODO:

 - [x] Admin utils to monitor & refresh gas
 - ~Admin util to withdraw unused gas~ (moved to https://github.com/oasisprotocol/demo-voting/issues/16)
 - [x] Multi-account multiplexing to balance transactions across gas paying accounts
 - [x] ~Remove 'publish votes' from the frontend?~ Reimplement publish votes (#15)

### CODE REVIEW

File: frontend/src/views/CreatePollView.vue [hunk 1]
```diff
@@ -1,7 +1,8 @@
 <script setup lang="ts">
 import { ref } from 'vue';
+import { ethers } from 'ethers';
 
-import { useDAOv1 } from '../contracts';
+import { useDAOv1, useGaslessVoting, useUnwrappedDAOv1, useUnwrappedGaslessVoting } from '../contracts';
```
Comment (matevz) [inline frontend/src/views/CreatePollView.vue:4]: Could you move this to a separate PR please? I would need that for my GSN voting as well and let's merge it separately ;)

File: frontend/src/contracts.ts [hunk 2]
```diff
@@ -25,10 +25,72 @@ export function useDAOv1(): ComputedRef<DAOv1> {
   });
 }
 
+export function useUnwrappedDAOv1(): ComputedRef<DAOv1> {
```
Comment (matevz) [inline frontend/src/contracts.ts:28]: Could you move `useUnwrappedDAO` and `useUnwrappedPollACLv1` to a separate PR please? I would need that for my GSN voting as well and let's merge it separately ;)

**Review by matevz — DISMISSED** [review]
First pass, I didn't test it yet, will do it later.

I suggest we have demo-voting supporting all three mechanisms:
- payable votes (current master branch)
- GSN votes
- gasless votes (we should come up with a fancy Oasis name for it?)

That being said I would not merge this until we cleanup the backend/contracts:
- EIP155Signer.sol should become part of @oasisprotocol/sapphire-contracts
- IERC165 should just be included from openzeppelin?
- RLPWriter should be part of @oasisprotocol/sapphire-contracts
- DAOv1 should be an interface with three implementations mentioned above

If we don't make it for EthCC, it's even fine to have it in a separate PRs for now.

Thoughts?

**Comment by CedarMist** [comment]
Good idea to move the EIP155Signer and RLPWriter to sapphire-contracts.

The unwrapped providers are kinda messy, could have an extra PR for them or
could just have 3 branches and copy what you need.

Would prefer to make finishing touches for deployment to mainnet on my
branch.

**Comment by CedarMist** [comment]
> DAOv1 should be an interface with three implementations mentioned above

There is no need for DAOv1 to be an interface with multiple implementations, unless we follow OpenZeppelin Governance interfaces or similar.

GSN implementation can use the same pattern as I did with GaslessVoting.

 - DAOv1 is assigned an optional 'proxy voter' contract, aka a trusted verifier which can override `msg.sender`
 - GSN contract interfaces with DAOv1 via `proxyVote` interface
 - Client detects which proxy voter implementation is being used, via ERC-165

**Comment by CedarMist** [comment]
I can't merge this because it's my branch. You will need to approve it.

**Review by matevz — APPROVED** [review]

**Pull request merged**